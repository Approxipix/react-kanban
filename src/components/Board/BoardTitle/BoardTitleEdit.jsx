import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { editBoardTitle } from '../../../redux/boardReducer/actions';
import ClickOutside from '../../ClickOutside';
import { Input, } from '../../BaseComponent';
import styled from 'styled-components'

const Form = styled.form`
  max-width: calc(100vw - 12rem);
`;

class BoardTitleEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.boardTitle || '',
    };
  }

  handleChange = (value) => {
    this.setState({
      title: value,
    });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    if (!title)  return;
    this.props.actions.editBoardTitle({
      boardId: this.props.boardId,
      boardTitle: title
    });
    this.props.toggleEditTitle()
  };

  render() {
    const { title } = this.state;
    let inputWidth = (title.length + 1) * 13;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditTitle}>
        <Form onSubmit={this.handleSubmit}>
          <Input
            type="text"
            size={'1.3rem'}
            weight={'600'}
            width={inputWidth + 'px'}
            value={title}
            placeholder="Edit board title"
            onChange={(e) => this.handleChange(e.target.value)}
            onBlur={this.handleSubmit}
            onKeyPress={(e) => inputWidth = ((e.target.value.length + 1) * 13)}
            autoFocus
            spellCheck={false}
          />
        </Form>
      </ClickOutside>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editBoardTitle: editBoardTitle,
    }, dispatch)
  };
}


export default connect(null, mapDispatchToProps)(BoardTitleEdit);
