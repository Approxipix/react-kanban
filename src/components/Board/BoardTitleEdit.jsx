import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { editBoardTitle } from '../../redux/boardReducer/actions';
import styled from 'styled-components'
import ClickOutside from '../ClickOutside';

const Form = styled.form`
  width: 25rem;
`;

const Input = styled.input`
  width: 100%;
  padding: .3rem .5rem;
  font-size: 1.3rem;
  color: #40424b;
  font-weight: 600;
  border: none;
  box-shadow: none;
`;

class BoardTitleEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.board.title || '',
    };
  }

  handleChange = (key, value) => {
    this.setState({
      ...this.state,
      [key]: value,
    });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    if (!title)  return;
    this.props.actions.editBoardTitle({ title: title });
    this.props.toggleEditing()
  };

  render() {
    const { title } = this.state;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditing}>
        <Form onSubmit={this.handleSubmit} id="board-edit-form">
          <Input
            autoFocus
            type="text"
            placeholder="Edit board title"
            value={title}
            onChange={(e) => this.handleChange('title', e.target.value)}
            spellCheck={false}
          />
        </Form>
      </ClickOutside>
    );
  };
}

function mapStateToProps(state) {
  const { rootReducer } = state;
  return {
    board: rootReducer.boards[rootReducer.currentBoardIndex],
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editBoardTitle: editBoardTitle,
    }, dispatch)
  };
}


export default connect(mapStateToProps, mapDispatchToProps)(BoardTitleEdit);
