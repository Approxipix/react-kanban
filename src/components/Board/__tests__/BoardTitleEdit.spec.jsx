import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import BoardTitleEdit from '../BoardTitle/BoardTitleEdit';
import configureStore from "redux-mock-store";

const mockStore = configureStore([]);

describe('<BoardTitleEdit>', () => {
  let store;
  let props;
  let component;
  let mockEditCardTitle = jest.fn();
  let mockToggleEditTitle = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();

    store = mockStore({});
    store.dispatch = mockEditCardTitle;
    props = {
      boardId: 'BoardID',
      boardTitle: 'Board Title',
      toggleEditTitle: mockToggleEditTitle
    };

    const wrapper = shallow(<BoardTitleEdit store={store} {...props} />);
    component = wrapper.find('BoardTitleEdit').dive();
  });

  it('should render with given state from Redux store', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should call toggleEditTitle if key code equal 27", () => {
    component.find('Input').simulate('keydown', { keyCode: 27 });

    expect(mockToggleEditTitle).toHaveBeenCalledTimes(1);
  });

  it('should call handleChange on title change with the correct params', () => {
    const spy = jest.spyOn(component.instance(), "handleChange");
    const mockEvent = {
      target: {
        name: "title",
        value: "Test title"
      }
    };

    component.find("Input").simulate("change", mockEvent);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(mockEvent);
  });

  it('should call setState on title', () => {
    const mockEvent = {
      target: {
        name: "title",
        value: "Test Title"
      }
    };
    const expectedState = {
      title: "Test Title",
    };

    component.instance().handleChange(mockEvent);

    expect(component.state()).toEqual(expectedState)
  });

  it("should call handleSubmit with the correct params", () => {
    const spy = jest.spyOn(component.instance(), "handleSubmit");
    const mockEvent = {
      preventDefault: jest.fn()
    };

    component.setState({ title: "Test Title" });
    component.instance().handleSubmit(mockEvent);

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should return handleSubmit if title is empty", () => {
    const spy = jest.spyOn(component.instance(), "handleSubmit");
    const mockEvent = {
      preventDefault: jest.fn()
    };

    component.instance().handleSubmit(mockEvent);

    expect(spy).toReturn();
  });

  it("should dispatch an action after handleSubmit", () => {
    const expectedAction = {
      type: 'EDIT_BOARD_TITLE',
      payload: {
        boardTitle: 'Test Title',
        boardId: 'BoardID',
      }
    };

    component.setState({ title: "Test Title" });
    component.instance().handleSubmit({ preventDefault: jest.fn() });

    expect(mockEditCardTitle).toHaveBeenCalledTimes(1);
    expect(mockEditCardTitle).toHaveBeenCalledWith(expectedAction);
  });

  it("should call toggleEditTitle after submit", () => {
    component.setState({ title: "Test Title" });
    component.instance().handleSubmit({ preventDefault: jest.fn() });

    expect(mockToggleEditTitle).toHaveBeenCalledTimes(1);
  });
});
