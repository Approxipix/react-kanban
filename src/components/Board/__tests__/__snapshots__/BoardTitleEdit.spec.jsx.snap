// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardTitleEdit> should render with given state from Redux store 1`] = `
<ClickOutside
  toggleOpened={[MockFunction]}
>
  <styled.form
    onSubmit={[Function]}
  >
    <Input
      autoFocus={true}
      margin="0"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      placeholder="Edit board title"
      size="1.3rem"
      spellCheck={false}
      type="text"
      value="Board Title"
      weight="600"
      width="156px"
    />
  </styled.form>
</ClickOutside>
`;
